src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,344 @@
+                {
+                    continue;
+                }
+
+                return false;
+            }
+
+            return true;
+        }
+
+        #region Parsing (public)
+        public static IPNetwork Parse(string s)
+        {
+            ArgumentNullException.ThrowIfNull(s);
+            return Parse(s.AsSpan());
+        }
+        public static IPNetwork Parse(ReadOnlySpan<char> s)
+        {
+            if (TryParseInternal(s, out var result, out var error))
+            {
+                return result;
+            }
+
+            throw new FormatException(error);
+        }
+
+        public static bool TryParse(string? s, out IPNetwork result)
+        {
+            if (s == null)
+            {
+                result = default;
+                return false;
+            }
+
+            return TryParseInternal(s.AsSpan(), out result, out _);
+        }
+
+        public static bool TryParse(ReadOnlySpan<char> s, out IPNetwork result)
+        {
+            return TryParseInternal(s, out result, out _);
+        }
+        #endregion
+
+        #region Private methods
+        private static bool TryParseInternal(ReadOnlySpan<char> s, out IPNetwork result, [NotNullWhen(false)] out string? error)
+        {
+            const int maxCharsCountAfterIpAddress = 4;
+            const int minCharsRequired = 4;
+
+            if (s.Length < minCharsRequired)
+            {
+                error = SR.dns_bad_ip_network;
+                result = default;
+                return false;
+            }
+
+            int separatorExpectedStartingIndex = s.Length - maxCharsCountAfterIpAddress;
+            int separatorIndex = s
+                .Slice(separatorExpectedStartingIndex)
+                .IndexOf(addressAndPrefixLengthSeparator);
+            if (separatorIndex != -1)
+            {
+                separatorIndex += separatorExpectedStartingIndex;
+
+                var ipAddressSpan = s.Slice(0, separatorIndex);
+                var prefixLengthSpan = s.Slice(separatorIndex + 1);
+
+                if (IPAddress.TryParse(ipAddressSpan, out var ipAddress) && int.TryParse(prefixLengthSpan, out var prefixLength))
+                {
+                    result = new IPNetwork
+                    {
+                        BaseAddress = ipAddress,
+                        PrefixLength = prefixLength
+                    };
+
+                    error = result.Validate()?.ParseExceptionMessage;
+                    return error == null;
+                }
+                else
+                {
+                    error = SR.dns_bad_ip_network;
+                    result = default;
+                    return false;
+                }
+            }
+            else
+            {
+                error = SR.dns_bad_ip_network;
+                result = default;
+                return false;
+            }
+        }
+
+        private readonly record struct ValidationError(
```

Comment:
Please don't use records here.  They add a lot of additional code we don't need.